scenarios/smart_trailer/digital_twin_providers/common/src/utils.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+use invehicle_stack_interfaces::service_discovery::core::v1::service_registry_client::ServiceRegistryClient;
+use invehicle_stack_interfaces::service_discovery::core::v1::DiscoverRequest;
+use invehicle_stack_interfaces::invehicle_digital_twin::v1::invehicle_digital_twin_client::InvehicleDigitalTwinClient;
+use invehicle_stack_interfaces::invehicle_digital_twin::v1::{EndpointInfo, FindByIdRequest};
+use log::{debug, info};
+use tonic::{Request, Status};
+
+/// Use Chariott Service Discovery to discover a service.
+///
+/// # Arguments
+/// * `chariott_uri` - Chariott's URI.
+/// * `namespace` - The service's namespace.
+/// * `name` - The service's name.
+/// * `version` - The service's version.
+/// # `communication_kind` - The service's communication kind.
+/// # `communication_reference` - The service's communication reference.
+pub async fn discover_service_using_chariott(
+    chariott_uri: &str,
+    namespace: &str,
+    name: &str,
+    version: &str,
+    communication_kind: &str,
+    communication_reference: &str,
+) -> Result<String, Status> {
+    let mut client = ServiceRegistryClient::connect(chariott_uri.to_string())
+        .await
+        .map_err(|e| Status::internal(e.to_string()))?;
+
+    let request = Request::new(DiscoverRequest {
+        namespace: namespace.to_string(),
+        name: name.to_string(),
+        version: version.to_string(),
+    });
+
+    let response = client
+        .discover(request)
+        .await
+        .map_err(|error| Status::internal(error.to_string()))?;
+
+    let service = response.into_inner().service.ok_or_else(|| Status::not_found("Did not find a service in Chariott with namespace '{namespace}', name '{name}' and version {version}"))?;
+
+    if service.communication_kind != communication_kind
+        && service.communication_reference != communication_reference
+    {
+        return Err(Status::not_found(
+            "Did not find a service in Chariott with namespace '{namespace}', name '{name}' and version {version} that has communication kind '{communication_kind} and communication_reference '{communication_reference}''",
+        ));
+    }
+
+    Ok(service.uri)
```

Comment:
Should this condition be an or?

```suggestion
    if service.communication_kind != communication_kind
        || service.communication_reference != communication_reference
    {
        Err(Status::not_found(
            "Did not find a service in Chariott with namespace '{namespace}', name '{name}' and version {version} that has communication kind '{communication_kind} and communication_reference '{communication_reference}''",
        ))
    } else {
        Ok(service.uri)
    }
```

Alternatively, it may be more intuitive to invert the if statement which should be an and condition

Reply:
Thanks, good catch. I will update this here. It also came from Ibeji, so I have filed this issue to change it there too: https://github.com/eclipse-ibeji/ibeji/issues/111